String_val: return 'const char *' when -safe-string is globally enabled

yallop opened the pull request:
(There is some discussion about this change in [Mantis PR 7594](https://caml.inria.fr/mantis/view.php?id=7594).)

#### Background & motivation

Since #687 it's possible to enable `-safe-string` globally at configure time.  #1252 makes configure-time `-safe-string` the default.

At present, `-safe-string` only affects OCaml code; C stubs that mutate OCaml strings will continue to compile without error. This may lead to subtle bugs, not least since the compiler now takes string immutability into account during optimizations (see, e.g., #703).

#### This change

This PR updates `String_val` to return `const char *` when `-safe-string` is enabled.  (When `-safe-string` is not enabled the type of `String_val` is unchanged.)  With typical settings, attempts to modify the result of `String_val` will result in a compiler warning, like this:

```
bar.c: In function ‘foo’:
/usr/lib/ocaml/caml/mlvalues.h:256:23: warning: initialization discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
 #define String_val(x) ((const char *) Bp_val(x))
                       ^
bar.c:6:13: note: in expansion of macro ‘String_val’
   char *p = String_val(s);
             ^~~~~~~~~~
```

This PR also adds a new macro `Bytes_val`, for accessing `bytes` values via `unsigned char *`.  Additionally, the types of several public functions have additional `const` qualifiers.

#### Compatibility
The change to `String_val` is slightly backwards-incompatible (but probably less so than defaulting to `-safe-string` (#1252)).  I ran some tests on `opam-repository`: of the 134 packages that use `Safe_string`, 4 will not compile with this patch if the compiler is configured with `-safe-string`.  ([More details in Mantis PR 7594](https://caml.inria.fr/mantis/view.php?id=7594#c18135).)

yallop committed:
String_val: return 'const char *' when -safe-string enabled.

yallop committed:
Add a Bytes_val macro.

yallop committed:
Add a changes entry for String_val return type change, Bytes_val.

mshinwell commented on byterun/extern.c:
```diff
@@ -480,7 +480,7 @@ static void extern_rec(value v)
 #endif
       }
-      writeblock(String_val(v), len);
+      writeblock((char *)String_val(v), len);
```
Would it be better to change `writeblock` to accept `const char*` instead?

yallop replied:
Yes, good idea.  I've made that change (8f27d9f).

mshinwell commented on byterun/str.c:
```diff
@@ -411,7 +411,7 @@ CAMLexport value caml_alloc_sprintf(const char * format, ...)
        so the size passed to vsnprintf is n+1. */
     va_start(args, format);
-    vsnprintf(String_val(res), n + 1, saved_format, args);
+    vsnprintf((char *)String_val(res), n + 1, saved_format, args);
```
The common pattern I'm looking for during review is that when you've got a `(char *)` cast, the thing being written into is a freshly-allocated string.  It looks like that is true---but maybe there should be a new `caml_alloc_writeable_string` function or similar which you could use instead, returning `(char*)`, to avoid these casts?

mshinwell replied:
(I mean returning the string, and the non-`const` pointer.)

yallop replied:
Do you have in mind something like this?
```C
value caml_alloc_writeable_string(mlsize_t len, char **p);
```
If so, that seems less clear to me, as returning multiple values in C is awkward:
```C
char *p;
value v = caml_alloc_writeable_string(len, &p);
vsnprintf(p, …)
```
But addressing the allocate + initialize pattern is a good idea.  I've added a new function `caml_alloc_initialized_string` (fd7fb761) that addresses the majority of these cases, where the string is initialized from a length and a source address.  Here's a typical change, from `byterun/str.c`:
```diff
-    res = caml_alloc_string(n);
-    memcpy((char *)String_val(res), buf, n);
+    res = caml_alloc_initialized_string(n, buf);
```

With this new function the length, which was previously passed to both `caml_alloc_string` and `memcpy`, is no longer repeated; there's no cast or call to `String_val` in the caller, and the caller never needs to touch a writeable string.  What do you think?

mshinwell replied:
Yeah, this seems good.

yallop committed:
Add caml_alloc_initialized_string.

Replace caml_alloc_string + String_val + cast + memcpy in a few places.

mshinwell commented:
Ah, I realised that we've forgotten one thing: updating the C interface section of the manual.  Could you do that?  OK with that change.

yallop committed:
Document the String_val -safe-string change, and Bytes_val.

yallop committed:
Document caml_alloc_initialized_string.

yallop commented:
I've updated the documentation for `String_val` and added documentation for `Bytes_val` in f1d1679, and documented `caml_alloc_initialized_string` in 18f83de.

damiendoligez commented:
> Since #687 it's possible to enable `-safe-string` globally at configure time. #1252 makes configure-time `-safe-string` the default.

Note that this is not entirely true. The configure-time `-safe-string` of #687 prevents any use of `-unsafe-string` altogether, while #1252 still leaves the option of compiling some files (or libraries) with `-unsafe-string`.

In your changes entry, you should probably write "at configure time" instead of "globally".

yallop commented:
Thanks for the clarification, @damiendoligez.  #1279 makes the change you suggest.

gasche commented:
As reported in [MPR#7658](https://caml.inria.fr/mantis/view.php?id=7658), the spacetime runtime needs to be updated after this change -- cc @mshinwell.
